File: apps/reading/src/api/recursiveTwinSessionPackCompose.ts
/**
 * Recursive twin session pack compose (pure).
 *
 * Operator vision: every information asset has a twin of insights + questions
 * (LLM as perfect note-taker). This pure layer packs caller-supplied twin
 * substrate signals for a session so they can be merged, referenced, and
 * searched — without writing the twin store.
 *
 * twin_store_mutated is always false.
 */

export interface TwinSessionMember {
  asset_id: string;
  twin_bound: boolean;
  insights: string[];
  questions: string[];
  /** Optional search hits already known (caller-supplied). */
  search_hits?: number | null;
}

export interface RecursiveTwinSessionPack {
  session_id: string;
  asset_ids: string[];
  insight_count: number;
  question_count: number;
  bound_count: number;
  unbound_count: number;
  /** True when ≥1 bound twin and ≥1 insight or question (scaffold ok). */
  pack_ready: boolean;
  /** Always false — pure pack never mutates twin store. */
  twin_store_mutated: false;
  notes: string[];
  authority: "recursive_twin_session_pack_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

function requireStringList(value: unknown, name: string): string[] {
  if (!Array.isArray(value)) {
    throw new Error(`${name} must be an array`);
  }
  const out: string[] = [];
  for (let i = 0; i < value.length; i++) {
    const s = value[i];
    if (typeof s !== "string" || !s.trim()) {
      throw new Error(`${name}[${i}] must be a non-empty string`);
    }
    out.push(s.trim());
  }
  return out;
}

/**
 * Compose a session pack of twin insights/questions for merge/search.
 * Never invents twin content; never mutates store.
 */
export function composeRecursiveTwinSessionPack(input: {
  session_id: string;
  members: TwinSessionMember[];
}): RecursiveTwinSessionPack {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  const session_id = requireNonEmpty(input.session_id, "session_id");
  if (!Array.isArray(input.members) || input.members.length === 0) {
    throw new Error("members must be a non-empty array");
  }

  const notes: string[] = [
    "twin_store_mutated=false — session pack intent only",
    "insights/questions are caller-supplied only (no invent content)",
  ];

  const asset_ids: string[] = [];
  const seen = new Set<string>();
  let insight_count = 0;
  let question_count = 0;
  let bound_count = 0;
  let unbound_count = 0;

  for (let i = 0; i < input.members.length; i++) {
    const m = input.members[i];
    if (!m || typeof m !== "object") {
      throw new Error(`members[${i}] must be an object`);
    }
    const asset_id = requireNonEmpty(m.asset_id, `members[${i}].asset_id`);
    if (seen.has(asset_id)) {
      throw new Error(`duplicate asset_id in members: ${asset_id}`);
    }
    seen.add(asset_id);
    asset_ids.push(asset_id);

    if (typeof m.twin_bound !== "boolean") {
      throw new Error(`members[${i}].twin_bound must be an explicit boolean`);
    }
    if (m.twin_bound) bound_count += 1;
    else unbound_count += 1;

    const insights = requireStringList(m.insights, `members[${i}].insights`);
    const questions = requireStringList(m.questions, `members[${i}].questions`);
    insight_count += insights.length;
    question_count += questions.length;

    if (m.search_hits !== undefined && m.search_hits !== null) {
      if (
        typeof m.search_hits !== "number" ||
        !Number.isFinite(m.search_hits) ||
        m.search_hits < 0 ||
        !Number.isInteger(m.search_hits)
      ) {
        throw new Error(
          `members[${i}].search_hits must be non-negative integer or null`,
        );
      }
    }
  }

  if (insight_count === 0 && question_count === 0) {
    notes.push(
      "no insights/questions supplied — pack scaffold only (no invent content)",
    );
  } else {
    notes.push(
      `insights=${insight_count} questions=${question_count} caller-supplied only`,
    );
  }

  const pack_ready =
    bound_count >= 1 && (insight_count > 0 || question_count > 0);
  if (!pack_ready) {
    if (bound_count < 1) {
      notes.push("pack_ready=false — need ≥1 twin_bound member");
    } else {
      notes.push(
        "pack_ready=false — bound twins present but no insights/questions",
      );
    }
  } else {
    notes.push("pack_ready=true — substrate pack ready for merge/search intent");
  }
  notes.push("twin_store_mutated=false");

  return {
    session_id,
    asset_ids,
    insight_count,
    question_count,
    bound_count,
    unbound_count,
    pack_ready,
    twin_store_mutated: false,
    notes,
    authority: "recursive_twin_session_pack_compose_advisory",
  };
}

export function formatRecursiveTwinSessionPackSummary(
  p: RecursiveTwinSessionPack,
): string {
  return (
    `twin pack ${p.session_id} · assets=${p.asset_ids.length} · ` +
    `insights=${p.insight_count} questions=${p.question_count} · ` +
    `ready=${p.pack_ready} · twin_store_mutated=false`
  );
}

File: substrate/recursive_twin_session_pack_compose.py
"""Recursive twin session pack compose (pure, advisory).

Packs caller-supplied twin insights/questions for a session.
twin_store_mutated is always False.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any


class RecursiveTwinSessionPackComposeError(ValueError):
    """Fail-closed validation for twin session pack compose."""


@dataclass(frozen=True)
class RecursiveTwinSessionPack:
    session_id: str
    asset_ids: tuple[str, ...]
    insight_count: int
    question_count: int
    bound_count: int
    unbound_count: int
    pack_ready: bool
    twin_store_mutated: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "session_id": self.session_id,
            "asset_ids": list(self.asset_ids),
            "insight_count": self.insight_count,
            "question_count": self.question_count,
            "bound_count": self.bound_count,
            "unbound_count": self.unbound_count,
            "pack_ready": self.pack_ready,
            "twin_store_mutated": False,
            "notes": list(self.notes),
            "authority": "recursive_twin_session_pack_compose_advisory",
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise RecursiveTwinSessionPackComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def _require_string_list(value: object, *, field: str) -> list[str]:
    if not isinstance(value, list):
        raise RecursiveTwinSessionPackComposeError(f"{field} must be an array")
    out: list[str] = []
    for i, s in enumerate(value):
        if not isinstance(s, str) or not s.strip():
            raise RecursiveTwinSessionPackComposeError(
                f"{field}[{i}] must be a non-empty string"
            )
        out.append(s.strip())
    return out


def compose_recursive_twin_session_pack(
    *,
    session_id: object,
    members: object,
) -> RecursiveTwinSessionPack:
    """Compose session twin pack. Never mutates twin store."""
    sid = _require_nonempty(session_id, field="session_id")
    if not isinstance(members, list) or len(members) == 0:
        raise RecursiveTwinSessionPackComposeError(
            "members must be a non-empty array"
        )

    notes: list[str] = [
        "twin_store_mutated=false — session pack intent only",
        "insights/questions are caller-supplied only (no invent content)",
    ]
    asset_ids: list[str] = []
    seen: set[str] = set()
    insight_count = 0
    question_count = 0
    bound_count = 0
    unbound_count = 0

    for i, m in enumerate(members):
        if not isinstance(m, dict):
            raise RecursiveTwinSessionPackComposeError(
                f"members[{i}] must be an object"
            )
        asset_id = _require_nonempty(
            m.get("asset_id"), field=f"members[{i}].asset_id"
        )
        if asset_id in seen:
            raise RecursiveTwinSessionPackComposeError(
                f"duplicate asset_id in members: {asset_id}"
            )
        seen.add(asset_id)
        asset_ids.append(asset_id)

        twin_bound = m.get("twin_bound")
        if not isinstance(twin_bound, bool):
            raise RecursiveTwinSessionPackComposeError(
                f"members[{i}].twin_bound must be an explicit boolean"
            )
        if twin_bound:
            bound_count += 1
        else:
            unbound_count += 1

        insights = _require_string_list(
            m.get("insights"), field=f"members[{i}].insights"
        )
        questions = _require_string_list(
            m.get("questions"), field=f"members[{i}].questions"
        )
        insight_count += len(insights)
        question_count += len(questions)

        search_hits = m.get("search_hits")
        if search_hits is not None:
            if (
                isinstance(search_hits, bool)
                or not isinstance(search_hits, int)
                or search_hits < 0
            ):
                raise RecursiveTwinSessionPackComposeError(
                    f"members[{i}].search_hits must be non-negative integer or null"
                )

    if insight_count == 0 and question_count == 0:
        notes.append(
            "no insights/questions supplied — pack scaffold only (no invent content)"
        )
    else:
        notes.append(
            f"insights={insight_count} questions={question_count} caller-supplied only"
        )

    pack_ready = bound_count >= 1 and (insight_count > 0 or question_count > 0)
    if not pack_ready:
        if bound_count < 1:
            notes.append("pack_ready=false — need ≥1 twin_bound member")
        else:
            notes.append(
                "pack_ready=false — bound twins present but no insights/questions"
            )
    else:
        notes.append(
            "pack_ready=true — substrate pack ready for merge/search intent"
        )
    notes.append("twin_store_mutated=false")

    return RecursiveTwinSessionPack(
        session_id=sid,
        asset_ids=tuple(asset_ids),
        insight_count=insight_count,
        question_count=question_count,
        bound_count=bound_count,
        unbound_count=unbound_count,
        pack_ready=pack_ready,
        twin_store_mutated=False,
        notes=tuple(notes),
        authority="recursive_twin_session_pack_compose_advisory",
    )


__all__ = [
    "RecursiveTwinSessionPack",
    "RecursiveTwinSessionPackComposeError",
    "compose_recursive_twin_session_pack",
]
